### PR TITLE
Fix stack overflow when finding blanket impls

### LINKED ISSUE #56701: cargo doc stack overflow (1.31)
The following code causes a stack overflow:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=df017106fc6a032b1b0959b1701b9781

Also here:
```
// compiling all this presents a stack overflow.  But removing any trait or impl fixes it

pub trait SimpleTrait {
    type SimpleT;
}

impl<Inner: SimpleTrait, Outer: Deref<Target = Inner>> SimpleTrait for Outer {
    type SimpleT = Inner::SimpleT;
}

pub trait AnotherTrait {
    type AnotherT;
}

impl<T, Simple: SimpleTrait<SimpleT = Vec<T>>> AnotherTrait for Simple {
    type AnotherT = T;
}

// this code block seems completely unrelated to the above, however Unrelated compiles alone, and SimpleTrait/AnotherTrait compile alone.
// But the combination won't compile.

pub struct Unrelated<Inner, UnrelatedT: DerefMut<Target = Vec<Inner>>>(UnrelatedT);

impl<Inner, UnrelatedT: DerefMut<Target = Vec<Inner>>> Deref for Unrelated<Inner, UnrelatedT> {
    type Target = Vec<Inner>;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}


pub fn main() { }
```

The output:
```
$ cargo doc
Documenting overflow-example v0.1.0 (/private/tmp/overflow_example)                                                                                      
                                                                                                                                                         
thread '<unknown>' has overflowed its stack                                                                                                               
fatal runtime error: stack overflow                                                                                                                       
error: Could not document `overflow-example`.                                                                                                             
 
Caused by:
 process didn't exit successfully: `rustdoc --edition=2018 --crate-name overflow_example src/main.rs --color always -o /private/tmp/overflow_example/target/doc -L dependency=/private/tmp/overflow_example/target/debug/deps` (signal: 6, SIGABRT: process abort signal)

$ cargo --version
cargo 1.31.0 (339d9f9c8 2018-11-16)

$ rustc --version
rustc 1.31.0 (abe02cefd 2018-12-04)
```

### PR DESCRIPTION
**Pull request by Aaron1011**
Currently, SelectionContext tries to prevent stack overflow by keeping
track of the current recursion depth. However, this depth tracking is
only used when performing normal section (which includes confirmation).
No such tracking is performed for evaluate_obligation_recursively, which
can allow a stack overflow to occur.

To fix this, this commit tracks the current predicate evaluation depth.
This is done separately from the existing obligation depth tracking:
an obligation overflow can occur across multiple calls to 'select' (e.g.
when fulfilling a trait), while a predicate evaluation overflow can only
happen as a result of a deep recursive call stack.

Fixes #56701

I've re-used `tcx.sess.recursion_limit` when checking for predication evaluation overflows. This is such a weird corner case that I don't believe it's necessary to have a separate setting controlling the maximum depth.

### CODE REVIEW

**Comment by Aaron1011** [comment]
@nikomatsakis: Are there any changes that you'd like me to make?

**Comment by nikomatsakis** [comment]
@Aaron1011 sorry, I haven't really reviewed this yet. I actually have it on my calendar to do later today (around 4pm if I stick to my schedule =) so I'll let you know! I'm trying hard to catch up, but conferences and other things have left me with quite a backlog. :(

File: src/librustc/traits/select.rs [hunk 1]
```diff
@@ -661,14 +665,15 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
         &mut self,
         stack: TraitObligationStackList<'o, 'tcx>,
         predicates: I,
+        recursion_depth: usize
```
Comment (nikomatsakis) [inline src/librustc/traits/select.rs]: can we call this the `evaluation_depth` or something like that?

It also feels like there should be a comment that links to the explanation below of why we have a *distinct* recursion depth tracking in use here.

File: src/librustc/traits/select.rs [hunk 2]
```diff
@@ -719,7 +742,8 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
                 obligation.cause.span,
             ) {
                 Some(obligations) => {
-                    self.evaluate_predicates_recursively(previous_stack, obligations.iter())
+                    self.evaluate_predicates_recursively(previous_stack, obligations.iter(),
```
Comment (nikomatsakis) [inline src/librustc/traits/select.rs]: Still, I feel like the *intention*  here was that the predicates we evaluate recursively would have a recursion depth that was incremented relative to `obligation.recursion_depth` -- but there are clearly some cases where that's not true. Perhaps *that* is really the bug?

**Review by nikomatsakis — COMMENTED** [review]
The code makes sense, but I can't escape the feeling that the bug is a failure to incremental the recursion depth in the recursive obligations, and not that we need a distinct recursion tracker for evaluation. Do you have ready access to the stack backtrace that causes the overflow in the example? (I can do a local build, too)

**Comment by Aaron1011** [comment]
@nikomatsakis: Here's the bottom part of the stack trace: https://gist.github.com/Aaron1011/6ee3bab4a34d6ef895cfb19d25bcd36c

File: src/librustc/traits/select.rs (hunk 2 above)
Comment (Aaron1011) [reply src/librustc/traits/select.rs]: I think you're right - it should be possible to increment the recursion depth of the obligations before passing them to `evaluate_predicates_recursively`, and check that recursion depth at the start of `evaluate_predicates_recursively`.

**Comment by Aaron1011** [comment]
@nikomatsakis: I've modified the PR to update the depth on nested obligations, instead of passing a separate `recursion_depth` parameter.

**Comment by Aaron1011** [comment]
It looks as though the new overflow check is getting hit before the old one, resulting in slightly worse error messages on overflow. I'll investigate more later today.

**Comment by nikomatsakis** [comment]
@Aaron1011 did you ever get a chance to investigate?

**Comment by Aaron1011** [comment]
@nikomatsakis: Sorry for the delay.

I ended up chaging the `recursion_limit` for several of the test. I believe these changes are justified, given that `recursion_limit` is now tracking additional steps (specifically, the interleaved `evaluate_predicate_recursively` calls:

* `src/test/run-pass/weird-exprs.rs`: I've increased `recursion_limit` from 128 to 256. This is necessary to keep `punch_card` compiling, due to the additional depth `evaluate_predicate_recursively`. Given that the `recursion_limit` attribute was needed in the first place, I don't think this is a huge issue.
* ` src/test/ui/did_you_mean/recursion_limit.stderr`: The error occurs earlier in the evaluation of the `Send` bounds. This is due to the additional depth from `evaluate_predicate_recursively` pushing the depth over the limit sooner. Since it's still essentially the same error, I think this is OK.
* `src/test/ui/error-codes/E0055.rs`: With this PR, a `recursion_limit` of 2 is no longer enough for compilation to reach `let foo_ref = ...` - overflow occurs while checking some standard library types. Given that such an incredibly low `recursion_limit` should never really be used in practice, I believe my adjustment is justified. Structurally, the code is essentially the same - it just uses a higher `recursion_limit` (and therefore greater number of chained references) to trigger the overflow.
* `src/test/ui/error-codes/E0275.stderr`: Overflow simply occurs slightly earlier in the expansion of the infinite type `Bar<Bar<...>>`. Since it's not actually possible to construct such a type, I don't see this an an issue.
`src/test/ui/issues/issue-20413.stderr`: This is the same as the previous case.

I don't think that's it's possible to avoid these UI changes without singificantly complicating the implementation. Since they only represent slight modifications to existing error messages, I think it should be fine.

File: src/librustc/traits/select.rs [hunk 3]
```diff
@@ -677,14 +678,25 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
...
+            Some(h) => self.check_recursion_limit(&obligation, h.obligation)?,
+            None => self.check_recursion_limit(&obligation, &obligation)?
+        }
 
         match obligation.predicate {
             ty::Predicate::Trait(ref t) => {
                 debug_assert!(!t.has_escaping_bound_vars());
-                let obligation = obligation.with(t.clone());
+                let mut obligation = obligation.with(t.clone());
+                obligation.recursion_depth += 1;
```
Comment (nikomatsakis) [inline src/librustc/traits/select.rs]: I don't think we need to increase the recursion depth here -- this is just translating the predicate from one form to another.

File: src/librustc/traits/select.rs [hunk 4]
```diff
@@ -693,8 +705,9 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
                 match self.infcx
                     .subtype_predicate(&obligation.cause, obligation.param_env, p)
                 {
-                    Some(Ok(InferOk { obligations, .. })) => {
-                        self.evaluate_predicates_recursively(previous_stack, &obligations)
+                    Some(Ok(InferOk { mut obligations, .. })) => {
+                        self.add_depth(obligations.iter_mut(), obligation.recursion_depth);
```
Comment (nikomatsakis) [inline src/librustc/traits/select.rs]: but it makes sense to increase the recursion depth *here* -- really, `subtype_predicate` should probably take the recursion depth as input, but I guess this is fine for now.

File: src/librustc/traits/select.rs [hunk 5]
```diff
@@ -708,8 +721,9 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
                 ty,
                 obligation.cause.span,
             ) {
-                Some(obligations) => {
-                    self.evaluate_predicates_recursively(previous_stack, obligations.iter())
+                Some(mut obligations) => {
+                    self.add_depth(obligations.iter_mut(), obligation.recursion_depth);
```
Comment (nikomatsakis) [inline src/librustc/traits/select.rs]: likewise here seems good

File: src/librustc/traits/select.rs [hunk 6]
```diff
@@ -813,10 +827,11 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
             ty::Predicate::Projection(ref data) => {
                 let project_obligation = obligation.with(data.clone());
                 match project::poly_project_and_unify_type(self, &project_obligation) {
-                    Ok(Some(subobligations)) => {
+                    Ok(Some(mut subobligations)) => {
+                        self.add_depth(subobligations.iter_mut(), obligation.recursion_depth);
```
Comment (nikomatsakis) [inline src/librustc/traits/select.rs]: and here

File: src/librustc/traits/select.rs [hunk 7]
```diff
@@ -1155,6 +1170,38 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
             .insert(trait_ref, WithDepNode::new(dep_node, result));
     }
 
+    // Due to caching of projection results, it's possible for a subobligation
```
Comment (nikomatsakis) [inline src/librustc/traits/select.rs]: I don't really think this is the reason why. It's more that things like `subtype` and `wf` don't take the "current recursion depth" as input and seem to just produce obligations at recursion depth 0, right? I would describe this function as a "workaround" for that scenario (which strikes me as odd but prob not worth changing in this PR)

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Nice -- I made a few suggestions, one of which *may* affect some of the tests in question.

File: src/librustc/traits/select.rs (hunk 3 above)
Comment (Aaron1011) [reply src/librustc/traits/select.rs]: I believe it is necessary here. Consider the following call stack:
1. `evaluate_predicate_recursively`
2. `evaluate_trait_predicate_recursively`
3. `evaluate_stack`
4. `evaluate_candidate`
5. `evaluate_predicates_recursively`
6. `evaluate_predicate_recursively`

Here, we've recursed back to `evaluate_predicate_recursively` without ever incrementing the recursion depth. To ensure that repeatedly recursing through the `ty::Predicate::Trait` match arm doesn't lead to a stack overflow, we need to increment the depth here.

File: src/librustc/traits/select.rs (hunk 7 above)
Comment (Aaron1011) [reply src/librustc/traits/select.rs]: I was referring to the `ty::Predicate::Projection` match arm, where we call `poly_project_and_unify_type`. If the projection result is cached, the subobligations will have an 'old' `recursion_depth`.

I'll update the comment to mention that several different factors create the need to ensure that our `recursion_depth` doesn't decrease.

**Comment by Aaron1011** [comment]
@nikomatsakis: Are there any other changes you'd like me to make?

**Comment by nikomatsakis** [comment]
@Aaron1011 will look shortly, trying to keep up with reviews, sorry =)

File: src/librustc/traits/select.rs (hunk 3 above)
Comment (nikomatsakis) [reply src/librustc/traits/select.rs]: I think that there should be an increase in between steps 4 and 5. In particular, the "nested obligations" returned by `evaluate_candidate` ought to be at a recursion level one higher than the candidate that spawned them.

**Review by nikomatsakis — COMMENTED** [review]
As I noted below, I still think this increment is unnecessary, and I'd prefer to remove it, as it is somewhat confusing (although I imagine it does no harm). If you think I'm wrong though, definitely let me know.

**Comment by nikomatsakis** [comment]
Er, huh, I can't figure out the GH display.

**Comment by Aaron1011** [comment]
@nikomatsakis: You were right! I've removed the uncessary increment (which also made some of the UI tests more closely resemble the current output).

**Review by nikomatsakis — APPROVED** [review]

**Pull request merged**